create_lmodrc.py

```diff
@@ -17,6 +17,81 @@
 }
 """
 
+GPU_LMOD_RC ="""require("strict")
+local hook = require("Hook")
+local open = io.open
+
+local function read_file(path)
+    local file = open(path, "rb") -- r read mode and b binary mode
+    if not file then return nil end
+    local content = file:read "*a" -- *a or *all reads the whole file
+    file:close()
+    return content
+end
+
+local function cuda_enabled_load_hook(t)
+    local frameStk  = require("FrameStk"):singleton()
+    local mt        = frameStk:mt()
+    local simpleName = string.match(t.modFullName, "(.-)/")
+    -- If we try to load CUDA itself, check if the full CUDA SDK was installed on the host in host_injections. 
+    -- This is required for end users to build additional CUDA software. If the full SDK isn't present, refuse
+    -- to load the CUDA module and print an informative message on how to set up GPU support for EESSI
+    if simpleName == 'CUDA' then
+        -- get the full host_injections path
+        local hostInjections = string.gsub(os.getenv('EESSI_SOFTWARE_PATH') or "", 'versions', 'host_injections')
+        -- build final path where the CUDA software should be installed
+        local cudaEasyBuildDir = hostInjections .. "/software/" .. t.modFullName .. "/easybuild"
+        local cudaDirExists = isDir(cudaEasyBuildDir)
+        if not cudaDirExists then
+            local advice = "but while the module file exists, the actual software is not entirely shipped with EESSI "
+            advice = advice .. "due to licencing. In order to be able to use the CUDA module, please follow the "
+            advice = advice .. "instructions available under https://www.eessi.io/docs/gpu/ \\n"
+            LmodError("\\nYou requested to load ", simpleName, " ", advice)
+        end
+    end
+    -- when loading CUDA enabled modules check if the necessary driver libraries are accessible to the EESSI linker,
+    -- otherwise, refuse to load the requested module and print error message
+    local haveGpu = mt:haveProperty(simpleName,"arch","gpu")
+    if haveGpu then
+        local arch = os.getenv("EESSI_CPU_FAMILY") or ""
+        local cudaVersionFile = "/cvmfs/pilot.eessi-hpc.org/host_injections/nvidia/" .. arch .. "/latest/cuda_version.txt"
+        local cudaDriverFile = "/cvmfs/pilot.eessi-hpc.org/host_injections/nvidia/" .. arch .. "/latest/libcuda.so"
+        local cudaDriverExists = isFile(cudaDriverFile)
```

Comment:
This seems not to be sufficient, if I start the container without a GPU the module still succeeds in loading. I suspect that `isFile` accepts that the link exists and ignores that it happens to be dead.

Reply:
Yes, that seems to be the case https://github.com/TACC/Lmod/blob/55d353b3a3a4649cbaaef18b2f00cdc017d7cc92/proj_mgmt/updateVersion#L126

Reply:
Hm, that's annoying. Should we revert back to your function and just make additional error codes pass? Any fix from Lmod will take a while, we defintily don't want to wait for that...

Reply:
I think it is ok to leave this as is, the case is a real corner one: you have the link available but you are on a node that doesn't have a GPU...no surprise that that won't work

Reply:
If the user does things "properly" that should not happen

Reply:
Well, I can see that corner case appearing if HPC sites point their `host_injections` variant symlink to a shared directory for convenience, and just put symlinks there to where the drivers are on their nodes that _do_ contain GPUs. But agreed, let's just go ahead for now. It's not like it breaks something that wasn't already broken: if you expect GPU software to work on a non-GPU node, you have different issues xD

Reply:
The only negative is that the error when you run the GPU code makes it look like all you have is a driver compatibility problem

Reply:
Oh ye if little faith! Lmod has been updated in the develop branch to address this.